type-check.ts

```diff
@@ -203,34 +218,78 @@ export function tcStmt(env : GlobalTypeEnv, locals : LocalTypeEnv, stmt : Stmt<n
  * @param value Type of the value passed into this destructure
  */
 function tcDestructure(env: GlobalTypeEnv, locals: LocalTypeEnv, destruct: Destructure<null>, value: Type): Destructure<Type> {
-  let values: Type[];
-  
-  if (destruct.isDestructured) {
-    throw new TypeCheckError("Destructured assignment not implemented");
-  } else if (destruct.targets.length === 1) {
-    values = [value];
+
+  if (!destruct.isDestructured) {
+    let target = destruct.targets[0];
+    return {
+      isDestructured: false,
+      targets: [{
+        target: tcAssignable(env, locals, target.target),
+        starred: target.starred,
+        ignore: target.ignore
+      }]
+    };
+  }
+
+  let types: Type[] = [];
+  if (value.tag === "class") { // This is a temporary hack to get destructuring working (reuse for tuples later?)
+    let cls = env.classes.get(value.name);
+    if (cls === undefined)
+      throw new Error(`Class ${value.name} not found in global environment. This is probably a parsing bug.`);
+    let attrs = cls[0];
+    attrs.forEach(val => types.push(val));
+  } else if (value.tag === "number" || value.tag === "bool" || value.tag === "none") {
+    types = [value];
   } else {
-    throw new TypeCheckError(`Non-destructured assignment must have 1 target. Got ${destruct.targets.length}`);
+    throw new TypeCheckError(`Type ${value.tag} cannot be destructured`);
   }
 
-  const tTargets = destruct.targets.map(({target, ignore, starred}, i) => {
-    const tValue = values[i];
+  function tcTarget(aTarget: AssignTarget<null>, valueType: Type): AssignTarget<Type> {
+    let {target, starred, ignore} = aTarget;
     const tTarget = tcAssignable(env, locals, target);
-    if (starred) {
-      throw new TypeCheckError("Starred values not supported")
-    }
     const targetType = tTarget.a;
-
-    if (!isAssignable(env, tValue, targetType)) {
-      throw new TypeCheckError(`Non-assignable types: Cannot assign ${tValue} to ${targetType}`);
-    }
-
+    if (!isAssignable(env, valueType, targetType))
+      throw new TypeCheckError(`Non-assignable types: Cannot assign ${valueType} to ${targetType}`);
     return {
       starred,
       ignore,
       target: tTarget,
     }
-  });
+  }
+
+  let tTargets: AssignTarget<Type>[] = [];
+  let starredIndex = -1;
+  for (let i = 0; i < destruct.targets.length; i++) {
+    let target = destruct.targets[i];
+    let valueType = types[i];
+    if (target.starred) {
```

Comment:
I would honestly just avoid trying to implement `starred` targets for right now. A lot of the implementation details rely on knowing how to parse out the target types, etc. which we currently don't really have. We can make enough of the examples work for thursday without this extra complexity yet.

Reply:
I've already added support for starred targets, so I would argue it's okay to leave them in.

Reply:
Fine. My guiding principles here are KISS and DRY. Following those, attempting to handle starred (other than a basic `throw new Error("not supported")`) feels superfluous when we don't have the initial spec of multiple assignment implemented yet.